### ISSUE TITLE
Eglot's dependencies state 2 different versions of project.el

### ISSUE THREAD

**Issue by gagbo** [joaotavora/eglot#492]
Related to #491  (Trying to find a clean solution)

I think the issue with installation (I'm using straight.el to install, through Doom) is that while `eglot` requires `package 0.3` :

```lisp
;; Version: 1.6
;; Author: João Távora <joaotavora@gmail.com>
;; Maintainer: João Távora <joaotavora@gmail.com>
;; URL: https://github.com/joaotavora/eglot
;; Keywords: convenience, languages
;; Package-Requires: ((emacs "26.1") (jsonrpc "1.0.9") (flymake "1.0.8") (project "0.3.0") (xref "1.0.1") (eldoc "1.0.0"))
```

`xref` at the `1.0.1` version is actually requiring the old version of `project`
```lisp
;;; xref.el --- Cross-referencing commands              -*-lexical-binding:t-*-

;; Copyright (C) 2014-2020 Free Software Foundation, Inc.
;; Version: 1.0.1
;; Package-Requires: ((emacs "26.3") (project "0.1.1"))
```

Dependency resolution is bound to have issues this way and I think it's the root of the problem. Somehow I ended up with `project 0.1.3` (I have no idea how Doom or straight handles these dependencies to be honest, but this is a separate issue at this point).

My point is that even if `package-install` appears to handle with the situation cleanly, it would be nice if we could avoid this situation to get out of "undefined behaviour" territory. But I don't know if this is possible. At least I want to bring out a possible source for the `(void function project-root)` issue.

**Comment by gagbo** [comment]
If having multiple package-requires just means "minimum versions" then we can possibly make that a straight.el issue I suppose

**Comment by joaotavora** [comment]
Thank you: this is very likely the culprit or _a_ culprit of the recent problems.  I do note that `package-install` does the right thing here. I mean `xref` does _require_ `project` v 0.1.1.  

> If having multiple package-requires just means "minimum versions" then we can possibly make that a straight.el issue I suppose

Not sure, but it might be yes.  In the meantime, let's ensure `xref.el` walks in tandem with `project.el`

**Comment by gagbo** [comment]
Even if we fix that, it looks like that the current version of straight pulls an old version of project.el anyway (found out why I was getting that `project.el @ 0.1.3`). I'll try to report that as well and see how it goes

**Comment by joaotavora** [comment]
> Even if we fix that, it looks like that the current version of straight pulls an old version of project.el anyway (found out why I was getting that project.el @ 0.1.3). I'll try to report that as well and see how it goes

In that case, looks like a straight out straight bug.  aha a yoke

**Comment by garyo** [comment]
Interesting. I'm using straight as well, and also have project.el@0.1.3. Straight is pulling it from https://github.com/emacsmirror/project which is that version.

But of course the version in emacs itself is https://github.com/emacs-mirror/emacs/blob/master/lisp/progmodes/project.el which is @0.3.0. And ELPA seems to be fine, http://elpa.gnu.org/packages/project.html.

**Comment by gagbo** [comment]
> In that case, looks like a straight out straight bug. aha a yoke

Indeed, it looks like the `gnu-elpa-mirror` source is not correctly plugged with straight (at least in my understanding), the discussion will probably move there (at least from a "end-user" point of view)

**Comment by eschulte** [comment]
On launching Eglot I'm getting an error "function `project-root` undefined."  I'm using project.el distributed with Emacs 26.3.  Could this be related to the same problem?  I can fix this problem by evaluating the following:
```lisp
(defun project-root (project) (car (project-roots project)))
```

**Comment by joaotavora** [comment]
> I'm using project.el distributed with Emacs 26.3. 

However, you should use a project.el distributed by GNU Elpa, and that is project.el 0.3.0.  In principle `package.el` will take care of that.  How did you install `eglot.el`?

```(defun project-root (project) (car (project-roots project)))

**Comment by gagbo** [comment]
For the record for Doom (using Straight.el to install packages) we ended up forcing the `gnu-elpa-mirror` recipe by hardcoding it https://github.com/hlissner/doom-emacs/pull/3291

**Comment by joaotavora** [comment]
But isn't this a `straight.el` bug?

**Comment by gagbo** [comment]
It probably is, but it wasn't moving fast enough with respect to the possible issues/bug reports that Doom would have to go through. You can see that I explained everything in the straight issue linked above ( 531 ), we'll see when raxod has time I think.

**Comment by joaotavora** [comment]
OK, thanks, I missed the link to the straight issue.

**Comment by eschulte** [comment]
Thanks for the quick reply, I just cloned eglot so that would explain why I don't have a recent enough project.el.  I'll update my project.el install from Elpa.

**Comment by joaotavora** [comment]
Yes, I use a clone copy of Eglot, too, but I use it under a master Emacs.  Otherwise, you have to use the package system, as complicated as it is occasionally.

**Comment by raxod502** [comment]
This problem should be fixed upstream at as per https://github.com/raxod502/straight.el/issues/531#issuecomment-643621955. I will take this opportunity to make two comments that are unrelated to the upstream issue:

* `straight.el` completely ignores version constraints and simply always installs the latest version, so that's not causing any trouble here.
* Unless I misunderstand, MELPA will ignore any version specification outside of the main package file `eglot.el`, so the version constraint for **all** files included in EGLOT is project>=0.3.0. While it might seem more semantically correct to include per-file dependencies, this doesn't actually work with `package.el` for a multi-file package. (It would work if EGLOT were distributed as multiple packages.)

**Comment by joaotavora** [comment]
Radon Rosborough <notifications@github.com> writes:

> This problem should be fixed upstream at as per
> raxod502/straight.el#531 (comment). I will take this opportunity to
> make two comments that are unrelated to the upstream issue:
>
> * straight.el completely ignores version constraints and simply always
> installs the latest version, so that's not causing any trouble here.
> 
> * Unless I misunderstand, MELPA will ignore any version specification
> outside of the main package file eglot.el, so the version constraint
> for all files included in EGLOT is project>=0.3.0. While it might seem
> more semantically correct to include per-file dependencies, this
> doesn't actually work with package.el for a multi-file package. (It
> would work if EGLOT were distributed as multiple packages.)

I'm not sure I understand 100% what you mean.  Can you summarize what
the previous two comments mean, in practice, for Eglot development?

If you're not updated on this, Eglot requires project 0.3 and xref 1.0.1
(which requires only the previous project 0.1.1).

Thanks,
João

**Comment by ivladak** [comment]
Sorry for jumping in late-ish. Just hit the problem that led me here and trying to get to the bottom of it.

> xref at the 1.0.1 version is actually requiring the old version of project

Why would this be a problem? According to the [docs](https://www.gnu.org/software/emacs/manual/html_node/elisp/Library-Headers.html#Library-Headers), in `Package-Requires`,
> The cadr of each sub-list is the *minimum* acceptable version number [...]

Thus, emacs should resolve the dependency with `package.el` whose version is at least as high as the maximum of these two (0.1.1, 0.3.0) requirements. Shouldn't it?

**Comment by gagbo** [comment]
I was just wondering if that was a problem or not. It makes sense that Package Requires has this kind of doc ; but I'm not sure that all package maintainer has a hard policy on backwards compatibility though.

Mostly the issue was that straight wasn't able to fetch/use a recent enough version of project, so the main discussion around this is found in the linked straight issue Straight.el cannot install a working (recent) version of Eglot raxod502/straight.el#531

**Comment by raxod502** [comment]
> I'm not sure I understand 100% what you mean.  Can you summarize what
> the previous two comments mean, in practice, for Eglot development?

Sorry. I misunderstood and thought that the two files quoted in the original post were both part of EGLOT, but they're not. My second comment is therefore not relevant.

My first comment is to point out that it's not possible for `straight.el` to have a problem with incorrect package version number resolution, even in principle, because `straight.el` completely ignores package version numbers, as if they do not exist. That is strictly a `package.el` thing.

**Comment by joaotavora** [comment]
> because straight.el completely ignores package version numbers, as if they do not exist. 

Again, I must ask you something, since this is confusing:

Eglot uses package version numbers.  It depends on minimum versions for some packages.  Eglot supports `package.el` only.  Just to confirm, are you saying I should advise Eglot users to NOT use `straight.el` because it ignores package version numbers?

In other words, I don't very much care about the implementation of `straight.el` right now. I just want to know if I should put in very bold letters that Eglot MUST be installed with `package.el`.

**Comment by raxod502** [comment]
> I just want to know if I should put in very bold letters that Eglot MUST be installed with package.el.

You should not do that.

**Comment by joaotavora** [comment]
> You should not do that.

So you're saying that even though `straight.el` "completely ignores package version numbers", it WILL ALWAYS correctly honour `eglot.el`'s request for proper versions of dependencies?  

Seems contradictory, but if you can guarantee that, I'm happy to close this issue and move on.  I'm sorry for the insistence, but this is the second straight/dependencies-related bug here, where I've had to invest some time.  It would be nice to really be sure.

**Comment by raxod502** [comment]
`straight.el` always installs the latest version of each package. By definition, this will satisfy your version constraints, will it not?

**Comment by joaotavora** [comment]
> By definition, this will satisfy your version constraints, will it not?

No, it won't, necessarily. Packages may break backward compatibility.  But they usually don't.  Let's hope they don't.

`package.el` also doesn't handle this, I think.

Thanks for clarifying. Let's close this issue.

**Comment by raxod502** [comment]
I mean, my point is just that `Package-Requires` doesn't _give_ you a way of defending against backwards compatibility problems. You can only specify a minimum version, not a maximum one. So in fact you don't really get anything from looking at the version numbers, aside from knowing if you need to upgrade. But assuming you're running the latest versions of everything, they don't tell you anything. That's why `straight.el` ignores the whole thing.

**Comment by joaotavora** [comment]
> That's why straight.el ignores the whole thing.

It doesn't outright ignore.  As you say, interpreting that information tells you if you need to upgrade or not. Though maybe straight _always_ upgrades for simplicity (and wastes time/bandwidth?).

Anyway, the way I read  "Package-Requires", it says: "package A needs a package B at a version compatible with V".  It's possible, maybe likely, that the set of package B's versions that are "compatible with V" may be specified in some other way in the future, perhaps using different  syntax in Package-Requires itself, or using some different data-source.  For now, I agree that everyone just assumes that any V1 >= V is "compatible with V".

**Comment by garyo** [comment]
With straight, you can "pin" package versions, and you can define "custom recipes" for packages, but unless you do those things, it _always_ installs the latest version of anything it installs. So `(straight-use-package 'eglot)` in emacs -Q will always install the latest version of eglot and all of its dependencies, transitively. And the same when you tell it to upgrade, via `(straight-pull-all)` and `(straight-check-all)`.
In fact, in my understanding at least, it gets you the head of the master branch rather than the latest tag. So it's often _more_ recent than the "current version" -- which of course is a double-edged sword.
